internal/querycoordv2/dist/dist_handler.go

```diff
@@ -181,120 +187,128 @@ func (dh *distHandler) updateSegmentsDistribution(ctx context.Context, resp *que
-}
-
-func (dh *distHandler) updateLeaderView(ctx context.Context, resp *querypb.GetDataDistributionResponse) {
-	updates := make([]*meta.LeaderView, 0, len(resp.GetLeaderViews()))
 
-	channels := lo.SliceToMap(resp.GetChannels(), func(channel *querypb.ChannelVersionInfo) (string, *querypb.ChannelVersionInfo) {
-		return channel.GetChannel(), channel
-	})
-
-	collectionsToSync := typeutil.NewUniqueSet()
-	for _, lview := range resp.GetLeaderViews() {
-		segments := make(map[int64]*meta.Segment)
-		for ID, position := range lview.GrowingSegments {
-			// To maintain compatibility with older versions of QueryNode,
-			// QueryCoord should neither process nor interact with L0 segments.
-			segmentInfo := dh.target.GetSealedSegment(ctx, lview.GetCollection(), ID, meta.CurrentTargetFirst)
-			if segmentInfo != nil && segmentInfo.GetLevel() == datapb.SegmentLevel_L0 {
-				continue
-			}
-			segments[ID] = &meta.Segment{
+		// Pre-allocate growing segments map
+		growings := lo.MapValues(lview.GetGrowingSegments(), func(position *msgpb.MsgPosition, id int64) *meta.Segment {
+			return &meta.Segment{
 				SegmentInfo: &datapb.SegmentInfo{
-					ID:            ID,
-					CollectionID:  lview.GetCollection(),
+					ID:            id,
+					CollectionID:  collectionID,
 					StartPosition: position,
-					InsertChannel: lview.GetChannel(),
+					InsertChannel: channelName,
 				},
-				Node: resp.NodeID,
+				Node: resp.GetNodeID(),
 			}
-		}
+		})
 
-		var version int64
-		channel, ok := channels[lview.GetChannel()]
-		if ok {
-			version = channel.GetVersion()
+		// Update DmChannel and register shard leader in same loop
+		dmChannel := &meta.DmChannel{
+			VchannelInfo: vChannelInfo,
+			Node:         resp.NodeID,
+			Version:      delegatorVersion,
+			View: &meta.LeaderView{
+				ID:                     resp.NodeID,
+				CollectionID:           collectionID,
+				Channel:                channelName,
+				Version:                delegatorVersion,
+				Segments:               lview.GetSegmentDist(),
+				GrowingSegments:        growings,
+				NumOfGrowingRows:       lview.GetNumOfGrowingRows(),
+				PartitionStatsVersions: lview.PartitionStatsVersions,
+				TargetVersion:          lview.GetTargetVersion(),
+				Status:                 lview.GetStatus(),
+			},
 		}
+		updates = append(updates, dmChannel)
 
-		view := &meta.LeaderView{
-			ID:                     resp.GetNodeID(),
-			CollectionID:           lview.GetCollection(),
-			Channel:                lview.GetChannel(),
-			Version:                version,
-			Segments:               lview.GetSegmentDist(),
-			GrowingSegments:        segments,
-			TargetVersion:          lview.TargetVersion,
-			NumOfGrowingRows:       lview.GetNumOfGrowingRows(),
-			PartitionStatsVersions: lview.PartitionStatsVersions,
+		serviceable := checkDelegatorServiceable(ctx, dh, dmChannel.View)
+		// trigger pull next target until shard leader is ready
+		if !serviceable {
+			dh.lastUpdateTs = 0
```

Comment:
lastUpdateTs is not guarded by lock

Reply:
each querynode has it's own dist_handler, and dist_handler pull data distribution in serial, the lock is unnecessary